src/observer.ts

```diff
@@ -8,31 +8,41 @@ export function useStaticRendering(enable: boolean) {
     isUsingStaticRendering = enable
 }
 
-export function observer<P>(baseComponent: FunctionComponent<P>): FunctionComponent<P> {
+export function observer<P extends object, TRef = {}>(
+    baseComponent: React.RefForwardingComponent<TRef, P>
+): React.MemoExoticComponent<
+    React.ForwardRefExoticComponent<React.PropsWithoutRef<P> & React.RefAttributes<TRef>>
+> {
     // The working of observer is explaind step by step in this talk: https://www.youtube.com/watch?v=cPF4iBedoF0&feature=youtu.be&t=1307
     if (isUsingStaticRendering) {
-        return baseComponent
+        return baseComponent as any
     }
 
     const baseComponentName = baseComponent.displayName || baseComponent.name
 
     // memo; we are not intested in deep updates
     // in props; we assume that if deep objects are changed,
     // this is in observables, which would have been tracked anyway
-    const memoComponent = memo((props: P) => {
-        const observerReaction = useObserverReaction(baseComponentName)
-
-        // render the original component, but have the
-        // reaction track the observables, so that rendering
-        // can be invalidated (see above) once a dependency changes
-        let rendering!: ReturnType<typeof baseComponent>
-        observerReaction.track(() => {
-            rendering = baseComponent(props)
+    const memoComponent = memo(
+        // we have to use forwardRef here because:
+        // 1. it cannot go before memo, only after it
+        // 2. forwardRef converts the function into an actual component, so we can't let the baseComponent do it
+        //    since it wouldn't be a callable function anymore
+        forwardRef((props: P, ref: any) => {
```

Comment:
I don't like the idea of having every observed component to be forwarding ref automatically. Can we like to figure how to do it optionally, eg. second argument to `observer` call? If not, then I would rather go with `forwardObserver` as a separate entity.

Official docs even mention it's not necessary that often and I personally have used it like once. It always fees like a hack if you need to communicate with a component directly.

> Ref forwarding is a technique for automatically passing a ref through a component to one of its children. This is typically not necessary for most components in the application.

Reply:
sounds good, done